File: BOJ7576.cpp
//
// Created by 전형진 on 2019-04-05.
//
#include <stdio.h>
#include <queue>
using namespace std;
int tom[1001][1001];
int visit[1001][1001];
int m,n;
int cnt =0;
int dx[]={0,1,0,-1};
int dy[]={1,0,-1,0};
queue<int> xq;
queue<int> yq;
void bfs(int y,int x) {


    //xq.push(x);
    //yq.push(y);
    visit[y][x] = 1;

    while (!yq.empty()) {
        int qs = yq.size();
        while (qs--) {
            int x = xq.front();
            int y = yq.front();
            xq.pop();
            yq.pop();
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && tom[ny][nx] == 0
                    && visit[ny][nx] == 0) {
                    //tom[ny][nx] = 3;
                    visit[ny][nx] = 1;
                    xq.push(nx);
                    yq.push(ny);
                }
            }
        }
        cnt++;
    }
}

int main(){
    scanf("%d %d",&m,&n);
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            scanf("%d",&tom[i][j]);
            if(tom[i][j] == 1){
                xq.push(j);
                yq.push(i);
            }

        }
    }

    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(tom[i][j] == 1)
                bfs(i,j);
        }
    }


    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(visit[i][j] == 0 && tom[i][j] == 0) {
                printf("-1");
                return 0;
            }

        }
    }
    printf("%d",cnt-1);
}

File: BOJ7576.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;
import java.util.StringTokenizer;

public class BOJ7576 {





    public static void main(String[] args) throws IOException {
       // Scanner sc = new Scanner(System.in);
        int m,n;
        int dx[] = {1,0,-1,0};
        int dy[] = {0,1,0,-1};
        int cnt = 0;
        boolean flag = false;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        Queue<tom> q = new LinkedList<>();
        int[][] map = new int[n+1][m+1];
        boolean[][] visit = new boolean[n+1][m+1];
        for (int i = 0; i < n; i++) {
            StringTokenizer st2 = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(st2.nextToken());
                if(map[i][j] == 1){
                    q.offer(new tom(i,j));
                }
            }
        }
        while(!q.isEmpty()) {
            int qs = q.size();
            while(qs>0) {
                int y = q.peek().y;
                int x = q.peek().x;
                visit[y][x] = true;
                q.poll();
                for (int i = 0; i < 4; i++) {
                    int nx = x + dx[i];
                    int ny = y + dy[i];

                    if (nx >= 0 && ny >= 0 && nx < m && ny < n && !visit[ny][nx]
                            && map[ny][nx] == 0) {
                        q.offer(new tom(ny, nx));
                        visit[ny][nx] = true;
                    }
                }
                qs--;
            }
            cnt++;
        }
        flag = true;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if(!visit[i][j] && map[i][j] == 0){
                    flag = false;

                }
            }
        }


        if(flag) System.out.println(cnt-1);
        else{
            System.out.println(-1);
        }

    }
    static class tom{
        int y;
        int x;

        public tom(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
}
